langchain/src/vectorstores/elasticsearch.ts

```diff
@@ -0,0 +1,212 @@
+import { Client, estypes } from "@elastic/elasticsearch";
+import { Embeddings } from "../embeddings/base.js";
+import { Document } from "../document.js";
+import { VectorStore } from "./base.js";
+
+type ElasticsearchKnnEngine = "hnsw";
+type ElasticsearchSimilarity = "l2_norm" | "dot_product" | "cosine";
+
+interface VectorSearchOptions {
+  readonly engine?: ElasticsearchKnnEngine;
+  readonly similarity?: ElasticsearchSimilarity;
+  readonly m?: number;
+  readonly efConstruction?: number;
+  readonly candidates?: number;
+}
+
+export interface ElasticsearchClientArgs {
+  readonly client: Client;
+  readonly indexName?: string;
+  readonly vectorSearchOptions?: VectorSearchOptions;
+}
+
+type ElasticsearchFilter = object;
+
+export class ElasticsearchVectorStore extends VectorStore {
```

Comment:
In Python this seems to be called ElasticVectorSearch: https://python.langchain.com/docs/modules/data_connection/vectorstores/integrations/elasticsearch

Not sure if you prefer the synchronization or if this is a better name now?